example/include/OMR/App/Object.hpp

```diff
@@ -0,0 +1,121 @@
+/*******************************************************************************
+ * Copyright (c) 2018, 2018 IBM Corp. and others
+ *
+ * This program and the accompanying materials are made available under
+ * the terms of the Eclipse Public License 2.0 which accompanies this
+ * distribution and is available at http://eclipse.org/legal/epl-2.0
+ * or the Apache License, Version 2.0 which accompanies this distribution
+ * and is available at https://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * This Source Code may also be made available under the following Secondary
+ * Licenses when the conditions for such availability set forth in the
+ * Eclipse Public License, v. 2.0 are satisfied: GNU General Public License,
+ * version 2 with the GNU Classpath Exception [1] and GNU General Public
+ * License, version 2 with the OpenJDK Assembly Exception [2].
+ *
+ * [1] https://www.gnu.org/software/classpath/license.html
+ * [2] http://openjdk.java.net/legal/assembly-exception.html
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-exception-2.0 OR LicenseRef-GPL-2.0 WITH Assembly-exception
+ *******************************************************************************/
+
+#if !defined(OMR_APP_OBJECT_HPP_)
+#define OMR_APP_OBJECT_HPP_
+
+#include <omrcfg.h>
+
+#include <cstdlib>
+#include <stdint.h>
+
+namespace OMR
+{
+namespace App
+{
+
+class ObjectSlotHandle
+{
+public:
+	ObjectSlotHandle(fomrobject_t* slot) : _slot(slot) {}
+
+	void setGcReference(fomrobject_t value) const { *_slot = value; }
+
+	fomrobject_t getGcReference() const { return *_slot; }
+
+private:
+	fomrobject_t* _slot;
+};
+
+class ObjectSlotIterator
+{
+public:
+	ObjectSlotIterator() {}
+
+	ObjectSlotIterator(const ObjectSlotIterator& other) : _slot(other._slot) {}
+
+	bool operator==(const ObjectSlotIterator& rhs) const { return _slot == rhs._slot; }
+
+	bool operator!=(const ObjectSlotIterator& rhs) const { return _slot != rhs._slot; }
+
+	ObjectSlotIterator& operator++()
+	{
+		++_slot;
+		return *this;
+	}
```

Comment:
This class is trying to implement an idiomatic c++ forward iterator, which normally uses references and overloaded operators.

My intention is to use this as:

```C++
for (ObjectSlotIterator it = object->begin(); it != object.end(); ++it) {
  ObjectSlotHandle slot = *it;
  visitor.edge(object, slot);
}
```

And I might as well mention, it works with range-based for loops in cxx11:

```c++
for(ObjectSlotHandle slot : object) {
  visitor.edge(object, slot);
}
```

Reply:
I ended up just using pointers as iterators, for the sake of simplicity.